src/v/cluster/seq_stm.cc

```diff
@@ -210,21 +78,23 @@ static bool is_sequence(int32_t last_seq, int32_t next_seq) {
 ss::future<checked<raft::replicate_result, kafka::error_code>>
 seq_stm::replicate(
   model::batch_identity bid,
-  model::record_batch_reader&& r,
+  model::record_batch_reader&& b,
   raft::replicate_options opts) {
     if (bid.has_idempotent() && bid.first_seq <= bid.last_seq) {
-        if (_insync_term != _c->term()) {
-            (void)ss::with_gate(_gate, [this] { return catchup(); });
-            return seastar::make_ready_future<
-              checked<raft::replicate_result, kafka::error_code>>(
+        // it's very wierd but moving the r-value from the arg
+        // to stack fixes a seg fault
+        auto br = std::move(b);
```

Comment:
I'm not fluent with this but when we do std::move we have performance / memory benefits and before this change on the path from produce via partition to consensus we were passing record_batch_reader only with move semantic so I feel uncomfortable to change that. Are you sure it's ok to pass record batch reader by value instead of `&&`?

Reply:
I think I wanted to use `auto&& br = std::move(b)` instead `auto br = std::move(b)` gonna try if it works.

Reply:
this is what you've got now: moving whatever reader `r` points to into the stack allocated record_batch_reader `r2` invokes the move assignment operator and then r2 is kept alive on the stack for the duration of the co-routine.

```
void replicate(model::record_batch_reader&& r) {
   auto r2 = std::move(r);
   next_method(std::move(r2));
}
```

what I'm saying is that the following is exactly the same. here you can think about `r` as being `r2` and all the parameters to a coroutine are kept alive.

```
void replicate(model::record_batch_reader r) {
   next_method(std::move(r));
}
```

nothing kept `&& r` alive because it isn't a value--it's just a pointer.

Reply:
> I think I wanted to use `auto&& br = std::move(b)` instead `auto br = std::move(b)` gna try if it works.

this is equivalent to before when only the input parameter was taken by r-value reference. remember that `T&&` is effectively the same as `T*` with some extra semantics. so `auto&& br = std::move(b)` would not be expected to alter the lifetime of the `b` input from the caller.

Reply:
I haven't tested `auto&&` but if it doesn't work, what do you think about the performance implications of passing `record_batch_reader instead` of `record_batch_reader&&`

Reply:
there is no cost concern: passing a `record_batch_reader&&` is cheap because it is equivalent to passing a pointer. but `record_batch_reader` uses pimpl pattern implemented with a `std::unique_ptr` so the cost of moving a record_batch_reader value is effectively identical. the difference is that the r-value reference communicate intent. but record_batch_reader is move-only anyway.